daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/InfosetImpl.scala

```diff
@@ -1450,37 +1449,40 @@ sealed class DIComplex(override val erd: ElementRuntimeData, val tunable: Daffod
    * When parsed, these arrays show up as separate DIArrays in the infoset. We need
    * to combine these separate arrays.
    * */
-  final def flattenAndValidateChildNodes(): Maybe[List[ElementRuntimeData]] = {
-    var ret: List[ElementRuntimeData] = List()
-    var groups = childNodes.groupBy(_.erd)
+  final def flattenAndValidateChildNodes(pstate: ParseOrUnparseState): Unit = {
+    val groups = childNodes.groupBy(_.erd)
+    childNodes.clear()
     groups foreach {
       case (erd, nodes) => {
         // Check min/maxOccurs validity while iterating over childNodes
-        val min0 = erd.minOccurs
-        val max0 = erd.maxOccurs
-        val isUnbounded = max0 == -1
+        val min = erd.minOccurs
+        val max = erd.maxOccurs
+        val isUnbounded = max == -1
 
         // Flatten multiple DIArrays into the first one
         if (erd.isArray) {
-          nodes.reduceLeft((a, b) => a.asInstanceOf[DIArray].concat(b.asInstanceOf[DIArray]))
+          val a = nodes(0).asInstanceOf[DIArray]
+          nodes.drop(1).foreach( b => a.concat(b.asInstanceOf[DIArray]))
           nodes.reduceToSize(1)
+
+          // Need to also remove duplicates from fastLookup
+          val fastSeq = nameToChildNodeLookup.get(a.namedQName)
+          if (fastSeq != null)
+            fastSeq.reduceToSize(1)
+        } else {
+          if (nodes.length > 1)
+            pstate.validationError("Multiple elements detected for scalar element %s", erd.namedQName)
```

Comment:
Maybe it would be helpful to mention that this is in an unordered sequence?

Also, the spec states this should be a processing error, not a validation error. We should have tests to make sure the PE is handled correctly.